### PR TITLE
Add ignore domains to Node SDK and ignore cage requests

### PR DESCRIPTION
**Pull request by davidnugent2425**
# Why
Gives user ability to pass in a list of domains they don't want to use with Outbound Relay. Requests to cages will also not be sent through Outbound Relay.

# How
```javascript
{ relay: true , ignoreDomains: ['httpbin.org, www.facebook.com'] }
// requests sent to urls such as https://httpbin.org/post or https://api.facebook.com will not be sent through Relay
```

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -37,17 +37,48 @@ class EvervaultClient {
...
+        ignoreEndsWith.push('@' + exact);
+    });
+
+    const _exactOrEndsWith = (domain, exactDomains, endsWithDomains) => {
+      if (exactDomains.includes(domain)) return true;
+      for (let end of endsWithDomains) {
+        if (domain?.endsWith(end)) return true;
+      }
+      return false;
+    };
```
Comment (shanecurran) [inline lib/index.js:61]:
```suggestion
  async _overloadHttpsModule(apiKey, tunnelHostname, ignoreDomains = []) {
    let pem = await this.http.getCert();

    const cagesHost = new URL(this.config.http.cageRunUrl).host;
    ignoreDomains.push(cagesHost);
    let ignoreExact = [];
    let ignoreEndsWith = [];
    ignoreDomains.forEach((domain) => {
        let exact = domain.startsWith('www.') ? domain.slice(4) : domain;
        ignoreExact.push(exact);
        ignoreEndsWith.push('.' + exact);
        ignoreEndsWith.push('@' + exact);
    });

    const _exactOrEndsWith = (domain, exactDomains, endsWithDomains) => {
      if (exactDomains.includes(domain)) return true;
      for (let end of endsWithDomains) {
        if (domain?.endsWith(end)) return true;
      }
      return false;
    };
```

File: README.md [hunk 2]
```diff
@@ -99,12 +99,22 @@ evervault.cagify(cageName: String, cageFunction: Function);
...
 ```javascript
 const evervaultClient = new Evervault('<API-KEY>', { relay: true });
 ```
 
+You may also optionally pass in an array of domains which you **don't** want to go through Relay, i.e. requests sent to these domains will not be decrypted.
+
+```javascript
+const evervaultClient = new Evervault(
+    '<API-KEY>',
+    { relay: true , ignoreDomains: ['httpbin.org, www.facebook.com'] }
```
Comment (shanecurran) [inline README.md:113]:
```suggestion
    { relay: true , ignoreDomains: ['httpbin.org', 'www.facebook.com'] }
```

**Comment by shanecurran** [comment]
Could this be refactored to have a single `isIgnored()` method which is called every time a request goes out?

Also instead of the manually `www.` splicing logic, it might make sense to use a third-party module to verify that a domain is in the same zone. Alternatively, it's probably acceptable to just only accept explicit hostnames. It's probably unlikely that developers will be calling a URL starting with `www.` considering they're mostly going to be APIs.

Other than that, approach looks good!

File: lib/index.js (hunk 1 above)
Comment (davidnugent2425) [reply lib/index.js:61]: Was it just the `if(!ignoreDomains...` line that was changed?

**Comment by davidnugent2425** [comment]
> Could this be refactored to have a single `isIgnored()` method which is called every time a request goes out?
> 
> Also instead of the manually `www.` splicing logic, it might make sense to use a third-party module to verify that a domain is in the same zone. Alternatively, it's probably acceptable to just only accept explicit hostnames. It's probably unlikely that developers will be calling a URL starting with `www.` considering they're mostly going to be APIs.
> 
> Other than that, approach looks good!

Refactored, the reason I did the url parsing manually is because users may want to just ignore `api.acme.com` and would want all `foo.api.acme.com` and `bar.api.acme.com` to be ignored while keeping `acme.com` non-ignored. I figured because it is only a couple of lines it would be better to just do it manually and know exactly what is being done rather than risking errors from a library. W.r.t the `www.` check, although it is a rare error, I figured it's better to add the check considering it doesn't add any extra lines to the code (Note: `'www.'.slice(4) === ''`, in case this was a worry). I'll take that example out of the README though

File: README.md [hunk 3]
```diff
@@ -99,12 +99,22 @@ evervault.cagify(cageName: String, cageFunction: Function);
...
 ```
 
+You may also optionally pass in an array of domains which you **don't** want to go through Relay, i.e. requests sent to these domains will not be decrypted.
+
+```javascript
+const evervaultClient = new Evervault(
+    '<API-KEY>',
+    { relay: true , ignoreDomains: ['httpbin.org', 'facebook.com'] }
+);
+// requests sent to urls such as https://httpbin.org/post or https://api.facebook.com will not be sent through Relay
```
Comment (shanecurran) [inline README.md:115]:
```suggestion
// Requests sent to URLs such as https://httpbin.org/post or https://api.facebook.com will not be sent through Relay
```

**Review by shanecurran — APPROVED** [review]

**Pull request merged**